dev/NavigationView/NavigationViewItemsFactory.cpp

```diff
@@ -35,34 +40,90 @@ winrt::UIElement NavigationViewItemsFactory::GetElementCore(winrt::ElementFactor
         return args.Data();
     }();
 
+    // Element is already of expected type, just return it
     if (auto const newItem = newContent.try_as<winrt::NavigationViewItemBase>())
     {
         return newItem;
     }
 
-    // Create a wrapping container for the data
-    auto nvi = winrt::make_self<NavigationViewItem>();
-    nvi->Content(newContent);
-    return *nvi;
-}
+    // Get or create a wrapping container for the data
+    auto const nvi = [this]() {
+        if (navigationViewItemPool.size() > 0)
+        {
+            auto nvi = navigationViewItemPool.back();
+            navigationViewItemPool.pop_back();
+            return nvi;
+        }
+        return winrt::make_self<NavigationViewItem>().try_as<winrt::NavigationViewItem>();
+    }();
 
-void NavigationViewItemsFactory::RecycleElementCore(winrt::ElementFactoryRecycleArgs const& args)
-{
+    auto const nviImpl = winrt::get_self<NavigationViewItem>(nvi);
```

Comment:
That's a good point. The issue here is that I need to return a strong reference to the NVI returned by `make_self` (otherwise the object will get freed). This is why returning a raw pointer (which is what `winrt::get_self` returns) to NavigationViewItem won't work, hence the added complexity.

Reply:
Hmm I see. So essentially, we are casting it to something to prevent it from getting freed once we are leaving the lambda? It seems a bit strange that we cast it back and forth just to get around the scope freeing, but if there is no other way to keep nvi a `const`, that is the way we have to do it. @StephenLPeters FYI.

Reply:
I think I am of the opinion that introducing a "needless" cast is too high a cost for maintaining this pattern that seems to be breaking down here. I think I would ditch the lambda.

---
In reply to: [444583286](https://github.com/microsoft/microsoft-ui-xaml/pull/2682#discussion_r444583286) [](ancestors = 444583286)